File: cpp/src/validation.cpp
#include "tide/ops.h"
#include "tide/kernel.h"
#include "tide/full.h"
#include "tide/aggregate.h"
#include "tide/read.h"
#include "tide/next.h"
#include "tide/region.h"
#include <algorithm>
#include <set>
#include <stdexcept>

namespace tide {
namespace {
void require(bool ok, const char* message) { if (!ok) throw std::invalid_argument(message); }
void check_tensor(const Tensor& x, const Tensor& reference, at::IntArrayRef shape) {
  require(x.defined() && x.device().is_cpu() && x.scalar_type() == reference.scalar_type()
          && x.sizes() == shape, "incompatible tensor dtype/device/shape");
  require(at::isfinite(x).all().item<bool>(), "nonfinite tensor");
}
}  // namespace
void validate_model(const Graph& g, const Model& m) {
  require(m.nodes.size() == g.nodes.size() && !m.nodes.empty(), "node weight count mismatch");
  const auto& ref = m.nodes[0].bias;
  require(ref.defined() && ref.dim() == 1 && ref.numel() > 0, "invalid model width");
  require(ref.scalar_type() == at::kFloat || ref.scalar_type() == at::kDouble, "FP32/FP64 required");
  const Index d = ref.numel();
  for (size_t node = 0; node < m.nodes.size(); ++node) {
    const auto& w = m.nodes[node];
    check_tensor(w.decay, ref, {d}); check_tensor(w.weight, ref, {d, d});
    check_tensor(w.bias, ref, {d}); check_tensor(w.read, ref, {d});
    require(static_cast<bool>(w.kernel), "state kernel is not configured");
    w.kernel->validate_policy(g.nodes[node], g.source_counts[node]);
    w.kernel->validate_weights(w);
    require(static_cast<bool>(w.read_kernel), "Read kernel is not configured");
    w.read_kernel->validate_weights(w);
    require(static_cast<bool>(w.next_kernel), "Next kernel is not configured");
    w.next_kernel->validate_weights(w);
    for (const auto& [name, value] : w.extra) check_tensor(value, ref, value.sizes());
    require(static_cast<bool>(w.full_kernel), "Full kernel is not configured");
    w.full_kernel->validate_weights(w, g.outgoing_ports.offsets[node+1] - g.outgoing_ports.offsets[node]);
    require(static_cast<bool>(w.aggregate_kernel), "Aggregate kernel is not configured");
    w.aggregate_kernel->validate_weights(w, g.source_counts[node]);
  }
  require(m.regions.size() == g.regions.size(), "region weight count mismatch");
  for (size_t r = 0; r < m.regions.size(); ++r) {
    const auto& w = m.regions[r];
    require(static_cast<bool>(w.kernel), "region kernel is not configured");
    for (const auto& [name, value] : w.extra) check_tensor(value, ref, value.sizes());
    w.kernel->validate_weights(w, region_layout(g, r));
  }
  require(m.input_scale.size() == g.inputs.size() && m.agg_scale.size() == g.edges.size()
          && m.edge_scale.size() == g.edges.size() && m.output_scale.size() == g.outputs.size(),
          "source scale count mismatch");
  for (const auto* group : {&m.input_scale, &m.agg_scale, &m.edge_scale, &m.output_scale})
    for (const auto& w : *group) check_tensor(w, ref, {});
}
std::vector<Atom> validate_window(const Graph& g, const Model& m, Continuation& q,
                                 const std::vector<External>& external, Index stop, Index seal) {
  require(q.identity == g.identity && q.batch_size > 0, "continuation identity/batch mismatch");
  require(q.cut >= 0 && q.cut <= stop && stop <= seal, "window is unsealed or precedes cut");
  const auto& ref = m.nodes[0].bias;
  const auto n = static_cast<Index>(g.nodes.size());
  const auto r = static_cast<Index>(g.regions.size());
  const auto p = static_cast<Index>(g.inputs.size());
  auto batch = [&q](Index b) { return b >= 0 && b < q.batch_size; };
  for (const auto& [owner, state] : q.states) {
    require(batch(owner.first) && owner.second >= 0 && owner.second < n, "invalid state owner");
    require(state.last_time < q.cut && state.last_time >= -1 && state.observations >= 0, "invalid state clock");
    check_tensor(state.value, ref, {m.width()});
    m.nodes[owner.second].kernel->validate_state(m.nodes[owner.second], state);
    for (const auto& [name, value] : state.slots) check_tensor(value, ref, value.sizes());
  }
  for (const auto& [owner, history] : q.history) {
    require(batch(owner.first) && owner.second >= 0 && owner.second < r, "invalid history owner");
    auto layout = region_layout(g, owner.second);
    validate_history(history, layout, ref, q.cut-1);
    m.regions[owner.second].kernel->validate_history(history, layout);
  }
  for (const auto& [owner, last] : q.ledger)
    require(batch(owner.first) && owner.second >= 0 && owner.second < p && last.first >= 0
            && last.second >= 0 && last.second < q.cut, "invalid input ledger");
  auto input = validate_external(g, m, q, external, stop, seal);
  std::set<std::tuple<Index, Index, Index>> seen;
  for (const auto& a : q.pending) {
    require(a.kind == 1 && a.source >= 0 && a.source < static_cast<Index>(g.edges.size()), "invalid pending edge");
    const auto& e = g.edges[a.source];
    require(batch(a.batch) && a.node == e.target && a.position >= 0 && a.position < q.cut
            && a.time >= q.cut && a.time >= a.position && a.time - a.position == e.delay,
            "invalid pending coordinate");
    require(seen.insert({a.batch, a.source, a.position}).second, "duplicate pending message");
    check_tensor(a.value, ref, {m.width()});
  }
  for (const auto& [owner, last] : input.ledger_updates) q.ledger[owner] = last;
  return input.atoms;
}
}  // namespace tide

File: python/tidegraph/validation.py
"""Fail early on malformed windows, states, tensor shapes and message identity."""
import torch
from .records import Atom


def validate_window(graph, model, continuation, external, stop, sealed_until):
    from .full import validate_program
    from .aggregate import validate_program as validate_aggregate
    offsets = graph.port_indexes[1].offsets
    incoming = graph.port_indexes[0].offsets
    for node, spec in enumerate(graph.nodes):
        validate_program(model.nodes[node], spec, offsets[node+1] - offsets[node])
        validate_aggregate(model.nodes[node], spec, incoming[node+1] - incoming[node])
    q = continuation
    if q.identity != graph.identity or q.batch_size < 1:
        raise ValueError("continuation identity or batch size mismatch")
    if not 0 <= q.cut <= stop <= sealed_until < 2**63:
        raise ValueError("window requires monotonic cut and explicit input seal")
    reference = model.nodes[0].bias
    def tensor(x):
        if x.device.type != "cpu" or x.dtype != reference.dtype or x.shape != (model.width,):
            raise ValueError("incompatible tensor device, dtype or shape")
        if not torch.isfinite(x).all():
            raise ValueError("nonfinite input/state")
    for (b, v), state in q.states.items():
        if not 0 <= b < q.batch_size or not 0 <= v < len(graph.nodes):
            raise ValueError("invalid state owner")
        if not -1 <= state.last_time < q.cut or state.observations < 0:
            raise ValueError("invalid state clock")
        tensor(state.value)
        model.nodes[v].validate(state)
        for value in state.slots.values():
            if value.device.type != "cpu" or value.dtype != reference.dtype or not torch.isfinite(value).all():
                raise ValueError("incompatible state slot dtype/device/value")
    for (b, r), counts in q.history.items():
        if not 0 <= b < q.batch_size or not 0 <= r < len(graph.regions):
            raise ValueError("invalid history owner")
        if any(v < 0 or v >= len(graph.nodes) or graph.nodes[v].region != r or c < 0
               for v, c in counts.items()):
            raise ValueError("invalid selector history")
    ledger = dict(q.ledger)
    for (b, p), (position, time) in ledger.items():
        if not (0 <= b < q.batch_size and 0 <= p < len(graph.inputs)
                and position >= 0 and 0 <= time < q.cut):
            raise ValueError("invalid input ledger")
    atoms = []
    for x in sorted(external, key=lambda e: (e.batch, e.port, e.position)):
        if not (0 <= x.batch < q.batch_size and 0 <= x.port < len(graph.inputs)
                and x.position >= 0 and q.cut <= x.time < stop):
            raise ValueError("invalid external coordinate")
        old_position, old_time = ledger.get((x.batch, x.port), (-1, -1))
        if x.position != old_position + 1 or x.time <= old_time:
            raise ValueError("port positions must be contiguous and times strictly increase")
        tensor(x.value)
        ledger[x.batch, x.port] = x.position, x.time
        atoms.append(Atom(x.batch, graph.inputs[x.port], x.time, 0, x.port, x.position, x.value))
    seen = set()
    for a in q.pending:
        if a.kind != 1 or not 0 <= a.source < len(graph.edges):
            raise ValueError("invalid pending edge")
        e = graph.edges[a.source]
        key = a.batch, a.source, a.position
        if (key in seen or not 0 <= a.batch < q.batch_size or a.node != e.target
                or not 0 <= a.position < q.cut or a.time != a.position + e.delay
                or not q.cut <= a.time < 2**63):
            raise ValueError("invalid pending message coordinate")
        seen.add(key)
        tensor(a.value)
    return atoms, ledger
